### ISSUE TITLE
Calico routing daemon couldn't start at baremetal server with k8s hosts in one cluster

### ISSUE THREAD

**Issue by f41gh7** [projectcalico/confd#288]
Hello, i have problem with my cluster after upgrading from 3.0.10 to 3.10.0 version.

 My setup is:
1) etcd cluster
2) kubernetes cluster with calico and etcd backend.
3) baremetal hosts with calico and etcd backend.

 After upgrading to 3.10.0 version - calico coulndt start routing daemon at baremetal nodes.

 Main reason is new Advertise IPs features. 

 Confd simply stacks at config generation, last line in confd log:
```2019-10-25 11:43:39.895 [INFO][41] watchersyncer.go 127: Sending status update Status=in-sync
2019-10-25 11:43:39.895 [DEBUG][41] client.go 331: Got status update: in-sync, syncer ready: false
2019-10-25 11:43:39.895 [INFO][41] client.go 350: Calico Syncer has indicated it is in sync
2019-10-25 11:43:39.895 [DEBUG][41] watchersyncer.go 241: Sending syncer updates (if any to send) NumUpdates=0
```

 So problem is at this wait, it can never be finished, because there is no incluster config.

https://github.com/projectcalico/confd/blob/release-v3.10/pkg/backends/calico/client.go#L367

  If `NewRouteGenerator` returns error, `c.OnInSync(SourceRouteGenerator)` will be never called and we are in trouble.

https://github.com/projectcalico/confd/blob/release-v3.10/pkg/backends/calico/client.go#L207 - there is only log message

 It happens because of new feauter was added:
https://github.com/projectcalico/confd/blob/release-v3.10/pkg/backends/calico/client.go#L189

 Possible solutions:
  Call c.OnInSync(SourceRouteGenerator) - if cannot create kubernetes cluster client.

**Comment by f41gh7** [comment]
more over, i got panic at confd client, when i`ve added service IP advertisment at k8s cluster.  

https://github.com/projectcalico/confd/blob/release-v3.10/pkg/backends/calico/routes.go#L290

 So, seems like i need option to disable Service/External IP routes syncing for non-k8s nodes.

**Comment by f41gh7** [comment]
updated MR, `c.rg = nil` must be set for correct behavior.

**Comment by caseydavenport** [comment]
It looks like we used to panic when we hit this error: https://github.com/projectcalico/confd/blob/release-v3.9/pkg/backends/calico/client.go#L180-L181

Is that the right behavior? Ultimately, this branch is hit when asking for service advertisement but when unable to talk to the Kubernetes API server, which is either a misconfiguration (like it sounds in this case) or a connection error or similar preventing access.

I don't see a great option... if we set OnInSync then we gloss over the problem and IPs won't get advertised. If we panic, then we crash and IPs won't get advertised. I guess if we do OnInSync then at least other function will continue to work.

**Comment by f41gh7** [comment]
First of all, i`ve tested it at 3.10 version. Second, i`ve mixed cluster with k8s/non k8s hosts.

 Panic is more complicated. Let me explain.

So,  after confd client started, we init router generator with constructor:
 https://github.com/projectcalico/confd/blob/release-v3.10/pkg/backends/calico/client.go#L199-L200

 It`s tricky, but newRouterGenerator constructor returns inited routegenerator, even if it cannot create kubernetes client.
 
 https://github.com/projectcalico/confd/blob/release-v3.10/pkg/backends/calico/routes.go#L55 

 Later, when someone add new serviceIPs, bgpsyncer calls OnUpdates function:

https://github.com/projectcalico/confd/blob/release-v3.10/pkg/backends/calico/client.go#L651

 There is check, if routergenerator != nil.
https://github.com/projectcalico/confd/blob/release-v3.10/pkg/backends/calico/client.go#L806

 But, it cannot be nil, because, it is always inited. And panic comes, after kubernetes client is called:
https://github.com/projectcalico/confd/blob/release-v3.10/pkg/backends/calico/routes.go#L290



 So, maybe it`s better to add type checking here:
https://github.com/projectcalico/confd/blob/release-v3.10/pkg/backends/calico/client.go#L200
 `if err == ErrNotInCluster {
                       c.OnInSync(SourceRouteGenerator)
			c.rg = nil
}
`

**Comment by f41gh7** [comment]
bump, any update?

**Comment by f41gh7** [comment]
currently, i`m using calico 3.10 with patched confd for 1 month.

**Comment by caseydavenport** [comment]
@f41gh7 sorry for the delay, I've just merged your PR and it will show up in Calico v3.11. Thanks!
